### PR TITLE
[Dev] Optimizer State and Master Weight Offloading

### PR DESCRIPTION
**Pull request by hxbai**
# What does this PR do ?
main branch: #2811 

### Summary

This PR introduces optimizer state offloading to CPU for the `DistributedOptimizer`, enabling significant GPU memory savings during training by temporarily moving optimizer states (exp_avg, exp_avg_sq) and master weights to CPU memory when not in use.

### Motivation

During the forward and backward passes, optimizer states occupy GPU memory but are not actively used. For large models, these states can consume a substantial portion of GPU memory. By offloading optimizer states to CPU after `optimizer.step()` and reloading them before the next step, we can reclaim this GPU memory for other operations like activation checkpointing or larger batch sizes.

### Performance & Memory Savings

**Memory Savings:**
- On **DeepSeek-V3**, this feature saves **15-20GB** of GPU memory (with 0.1-0.2s/iter overhead on GB200)

**Comparison with Optimizer CPU Offloading:**

| Aspect | CPU Offloading (ZeRO-Offload style) | State Offloading (This PR) |
|--------|-------------------------------------|----------------------------|
| **Where optimizer runs** | CPU | GPU |
| **D2H/H2D frequency** | Every step (gradients + params) | Every step (states only) |
| **Compute location** | Adam step on CPU | Adam step on GPU |
| **Best for** | Memory-constrained, bandwidth-limited | High-bandwidth interconnects (NVLink, GB200) |

More details:
- With higher H2D/D2H bandwidth such as **GB200**, the state offloading has significantly less overhead
- Async transfer overlapping
- Optimizer step still runs on GPU, avoiding CPU compute bottleneck
- Pinned memory enables maximum PCIe/NVLink bandwidth utilization
- Currently requires TE FusedAdam optimizer

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/b3f0ab327c14688f1b91cbfb0fe0957e26bbd8be

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/a27aa49741fe424969b1b8c912edd4d1a5421801

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/3ec46ba2bb3511ee1cd24033ce6b831bee18658d

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/d2e47737aa0ef47f8cbbf21887b324d7da19cf2d

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/39d5ba532f98a993552efe96828985401ba74dc8

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/cea43400bdd0bd06c07ba968129f0d2dedd2c6ba

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/cecd4fc765e87c3f1db58cc0596dd4c358096a30

**Review by yaox12 — APPROVED** [review]

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2760/commits/e3ab7ccba5bbc441e988d76da0145da50368c2d1
